### ISSUE TITLE
VSCode leaves several running processes after exit

### ISSUE THREAD

**Issue by dimitry-ishenko** [microsoft/vscode-jupyter#1626]
Issue Type: <b>Bug</b>

Steps to reproduce:

1. Launch vscode.
2. Open or create a Jupyter Notebook.
3. Do some work on it.
4. Close vscode.
5. Optional: Rinse and repeat.

Expected behavior:

Any processes started by vscode are closed.

Actual behavior:

After launching and closing vscode a few times (and realizing I am starting to run out of memory) I notice this:

```bash
user@laptop:~$ ps aux | grep -v networkd-dispatcher | grep -v unattended | grep -v grep | grep python
user   344513  0.2  0.6 4115828 102040 ?      Sl   May19   3:19 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   345452  0.4  5.3 5162548 858400 ?      Sl   May19   5:46 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   345607  0.2  0.6 4443632 102484 ?      Sl   May19   3:01 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   353204  0.2  0.6 3788220 105468 ?      Sl   May19   1:55 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   353605  0.0  0.2 117072 46508 ?        S    May19   0:01 /usr/bin/python3 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/pythonFiles/pyvsc-run-isolated.py vscode_datascience_helpers.daemon --daemon-module=vscode_datascience_helpers.jupyter_daemon -v
user   353771  0.3  0.6 3796468 105576 ?      Sl   May19   2:29 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   353933  0.2  0.2 3722708 39000 ?       Sl   May19   1:31 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   354840  0.2  0.2 3796424 35728 ?       Sl   00:35   1:34 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   355331  0.9  5.4 6179224 869260 ?      Sl   00:39   6:06 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   356104  0.3  0.6 3862044 111216 ?      Sl   00:52   1:52 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   366116  1.6  0.4 3796252 76032 ?       Sl   10:27   0:35 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   366803  2.3  0.4 1007044 69596 ?       Sl   10:32   0:44 /usr/bin/python3 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/pythonFiles/pyvsc-run-isolated.py vscode_datascience_helpers.daemon --daemon-module=vscode_datascience_helpers.jupyter_daemon -v
user   366808 10.1  5.4 5509256 871684 ?      Sl   10:32   3:16 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/languageServer.0.5.50/Microsoft.Python.LanguageServer
user   366831  0.0  0.2 551048 35328 ?        Ssl  10:32   0:01 /usr/bin/python3 -m ipykernel_launcher -f /home/user/.local/share/jupyter/runtime/kernel-33b1ca6c-b09c-41dd-b2fc-c466b2511581.json
user   366855  0.0  0.2 550640 35408 ?        Ssl  10:32   0:01 /usr/bin/python3 -m ipykernel_launcher -f /home/user/.local/share/jupyter/runtime/kernel-a5a7d55e-0ebc-4b4b-aac4-4d1dddff8450.json
user   368190  7.6  1.2 2471816 200372 ?      Ssl  10:45   1:27 /usr/bin/python3 -m ipykernel_launcher -f /home/user/.local/share/jupyter/runtime/kernel-b00685be-189e-4f07-80e0-237ab6772827.json
user   368212  0.0  0.2 550648 43728 ?        Ssl  10:45   0:00 /usr/bin/python3 -m ipykernel_launcher -f /home/user/.local/share/jupyter/runtime/kernel-918a11e0-f81e-4740-a3c1-e5d5ca074570.json
user   369055  4.4  1.3 2076996 219232 pts/0  Sl+  10:57   0:19 /usr/bin/python3 /home/user/.local/bin/ipython
user   369168  2.0  0.3 709300 62112 ?        Sl   10:58   0:08 /usr/bin/python3 /home/user/.vscode-oss/extensions/ms-python.python-2020.5.80290/pythonFiles/pyvsc-run-isolated.py vscode_datascience_helpers.daemon --daemon-module=vscode_datascience_helpers.jupyter_daemon -v
user@laptop:~$ 
```

Extension version: 2020.5.80290
VS Code version: Code - OSS 1.46.0 (3984852b33bd6897dd83108ac305c567d49e8fda, 2020-05-15T14:46:12.527Z)
OS version: Linux x64 5.4.0-29-generic

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz (8 x 1137)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: disabled_software<br>skia_renderer: disabled_off_ok<br>video_decode: unavailable_off<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off_ok<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|1, 2, 2|
|Memory (System)|15.23GB (0.93GB free)|
|Process Argv|--no-sandbox --unity-launch|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|ubuntu|
|XDG_CURRENT_DESKTOP|Unity|
|XDG_SESSION_DESKTOP|ubuntu|
|XDG_SESSION_TYPE|x11|
</details>

**Comment by greazer** [comment]
Old bug with no new comments for nearly a year. Closing.

**Comment by dimitry-ishenko** [comment]
It's still a bug. Never been handled. Why would you close it?

**Comment by dimitry-ishenko** [comment]
@greazer this was never fixed. I've observed it few weeks ago. Please reopen.

**Comment by dimitry-ishenko** [comment]
Thank you @rchiodo. I didn't want to keep bugging ya'll as I know everybody is busy. I am still patiently waiting for this to get fixed.

**Comment by JPvRiel** [comment]
I can confirm this behavior of leaving a stale python instance of `-m ipykernel_launcher` still happens with current versions of vscode on an up to date Ubuntu 20.04 OS:

```
code --version
1.59.0
379476f0e13988d90fab105c5c19e7abc8b1dea8
x64
```

And

```bash
$ code --list-extensions --show-versions | grep jupyter
ms-toolsai.jupyter@2021.8.1236758218
```

By the way `code -s` is useful to see the `ipykernel_launcher` instances started. Close vscode, and the PIDs get orphaned and picked up/owned by the systemd user session pid because vscode failed to reap those processes when closing.

To find details of stuff left behind/not cleaned up by vscode:

```bash
for p in $(pgrep -f ipykernel_launcher); do ps -f -p $p; done | more
```

**Comment by rchiodo** [comment]
The root cause of this bug is the daemon's we spin up to make kernel's launch faster.

I believe you can workaround the problem by eliminating these files here (copy them elsewhere first in case that breaks something else):

```
<extension dir>/pythonFiles/lib/python/pyls_jsonrpc
```

**Comment by JPvRiel** [comment]
> root cause of this bug is the daemon's we spin up to make kernel's launch faster.

Thanks @rchiodo 

So far, removing `pyls_jsonrpc` seems to work:

```bash
mkdir /tmp/ipy_buggy_launch
mv ~/.vscode/extensions/ms-toolsai.jupyter-2021.8.1236758218/pythonFiles/lib/python/pyls_jsonrpc /tmp/ipy_buggy_launch
```

Since doing this, haven't yet reproduced orphaned IPython instances after vscode closes.

**Comment by dimitry-ishenko** [comment]
@rchiodo @JPvRiel is there a measurable downside of doing this?

**Comment by rchiodo** [comment]
Downside is longer startup times for kernels. But that's relative. On some windows machines this can cost you 3-5 seconds for startup. On a linux box, it's a lot less time.

**Comment by seVenVo1d** [comment]
I am experiencing a similar problem. Here are my specs
```
Version: 1.61.2
Date: 2021-10-19T14:58:13.605Z
Electron: 13.5.1
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Linux x64 5.14.14-200.fc34.x86_64
```
and I am using Fedora 34. 

When I run a .ipynb code, I see these instances

![before](https://user-images.githubusercontent.com/45866787/139540648-c315dabe-02d9-4182-8938-6f15b0953a8f.png)
and after closing the VSCode

![after](https://user-images.githubusercontent.com/45866787/139540650-69d993a5-ffe1-42c5-9415-771f683efbc2.png)

So the VSCode and its related things are being closed in couple of seconds however the python instance still runs.

Details of the `python` process
![details](https://user-images.githubusercontent.com/45866787/139540655-006710cf-0eaf-4d72-893d-1ab47ff26a81.png)

Another major problem is that another' python' process appears after closing the VSCode when you run another .ipynb file. For example, I have run a .ipynb code, closed the VSCode, and rerun the same code. After doing this 4 times, I am seeing this.

![rerun](https://user-images.githubusercontent.com/45866787/139541105-7a43d2ad-2638-4487-afa1-b26d811e3eac.png)

As you can see, there are 4 .ipykernel_launcer instances, and each of them is taking about 100 Mb RAM, which I believe is a lot. I think this is a severe problem. Is it some sort of a bug, or its a problem on my side?

**Comment by rchiodo** [comment]
Perhaps if we change the daemon to start as a process group, it should kill the child process automatically.

**Comment by rchiodo** [comment]
I cannot seem to reproduce this anymore.

**Comment by rchiodo** [comment]
Oh no wait, I can repro if I close VS code

**Comment by DonJayamanne** [comment]
Thought I'd document what we discussed some time ago:
* Keep track of the PIDs started by daemon
	* Remove the PID from this list when a kernel dies (deliberate or otherwise)
	* When existing VS Code kill all processes with PIDs in the above list (previous step is necessary to ensure we don't kill processes we no longer own, if a PID is re-used after the original one is killed)

**Comment by rchiodo** [comment]
Oh you mean send the PID back to the extension host and have the extension host clean them up?

**Comment by DonJayamanne** [comment]
> Oh you mean send the PID back to the extension host and have the extension host clean them up?

Yes, I believe we already do send the PID back to the extension host (if not should be easy enough to do that).

**Comment by DonJayamanne** [comment]
Just checked, we don't send it back (instead we just log it).

**Comment by rchiodo** [comment]
I'm wondering why we need a daemon at all for kernel launch. What does it buy us? I thought the daemons were there in order to run python code without having to start another process but in this scenario, it launches the kernel as another process anyway.

**Comment by DonJayamanne** [comment]
This was a faster way to start the kernel process, specially with conda environments.
Let me check if this is still true.

**Comment by rchiodo** [comment]
Because the daemon already has the correct environment? But shouldn't our caching already have it too?

**Comment by DonJayamanne** [comment]
> he correct environment? But shouldn't our caching already have it too?

Yes, I think now that we cache the env variables for conda envrionments, that might be unnecessary.
I"m testing to confirm this locally, seems promising.

**Comment by DonJayamanne** [comment]
Yes, confirmed, its actually the same, even with a very slow conda env.
I'd say lets try and remove the whole daemon code.
Or we could just disable daemons as a first step (in settings) `"jupyter.disablePythonDaemon": true,`

Please could you try this at your end and confirm:
* The kernel processes don't hand ground (not related to whether we remove daemons or not, but lets try that either way)
* Startup is still fast (Its fast at my end, hence I think its safe to yank this feature)

**Comment by DonJayamanne** [comment]
Oh wait @rchiodo, when we try to run other code agains the pyhton envionment, then things can be slow.
As we start a whole new process and if I remember correctly its slow on windows.
Let me check on my (slow) windows PC

**Comment by rchiodo** [comment]
Yeah other code can use the daemon still. It's just raw kernel start that's the problem here.

**Comment by rchiodo** [comment]
> The kernel processes don't hand ground

Not sure what you meant to type here.

**Comment by rchiodo** [comment]
And disabling python daemons definitely fixes the problem.

**Comment by DonJayamanne** [comment]
Dont' `hang` around.

**Comment by DonJayamanne** [comment]
Tested on windows too and still fast, we're also caching whether ipykernel & the like are installed or not.
Hence we've reduced a lot of the python process launches, i.e. daemon doesn't provide much benefit at all - except when using an env for the very first time.

**Comment by DonJayamanne** [comment]
I'd suggest disabling this as a first step (via setting or code).

**Comment by rchiodo** [comment]
You think there might be cases where it's still necessary?

**Comment by IanMatthewHuff** [comment]
Marking this as verified though I don't have an exact repro to test here. Just working with insiders and making sure things are cleaned up after shutdown. And I'm not seeing anything left over.

**Comment by rchiodo** [comment]
Repro case is using linux/wsl, start a notebook kernel, close VS code itself.

It used to leave the kernel processes running.

**Comment by dimitry-ishenko** [comment]
How does one know if this has landed in the insiders build? I've still experienced this today, but my build was about a week old

**Comment by IanMatthewHuff** [comment]
Whoops my mistake. I'm on Mac, didn't notice the WSL / linux specificity. I'll remove my verification.

**Comment by dimitry-ishenko** [comment]
Thanks @rchiodo

**Comment by dimitry-ishenko** [comment]
@rchiodo I can confirm that this handles the problem. Sorry dunno how to add verified label... 🤷🏻‍♂️

**Comment by dimitry-ishenko** [comment]
@rchiodo any idea when this will land in the release build? Or is it safe to stay on pre-release?

**Comment by rchiodo** [comment]
Next week sometime it will be in stable. Pre-release is certainly less stable. It ships every night after running our tests.

**Comment by augaug** [comment]
Total noob here, but maybe there's something simple to look at in my JSON settings, what I'm observing is closing vscode leaves two conda processes up and running, one of which is zero but the other is consuming ~30% cpu. Basically the entire time I have vscode open there is a conda process at 30% regardless of what I'm doing. Then when I quit vscode and look at task manager it never really quits, those two conda processes are still there, one pegged at ~30%, the other at 0 and a Console Window host also at 0, and vscode is still visible as a running process in task manager, like it just won't quit. Any suggestions?

**Comment by rchiodo** [comment]
@augaug what is the command line for those processes? That might be conda discovery flaking out.

**Comment by augaug** [comment]
@rchiodo - how would I find that info? I can provide just not sure what you mean exactly. 

One thing I noticed that is slightly odd is that I choose my interpreter, it's a conda env, and it says "('arcgispro-py3-clone':conda)" in status bar of vscode which is correct...but then at some point it just switches, has the same interpreter path etc. and everything works, but no longer says (:conda), it says "Python 3,7,11 64-bit (system)" - I found that odd, not sure if it means anything.
